modyn/supervisor/internal/triggers/datadrifttrigger.py

```diff
@@ -36,34 +67,27 @@ def __init__(self, config: DataDriftTriggerConfig):
         self.encoder_downloader: EmbeddingEncoderDownloader | None = None
         self.embedding_encoder: EmbeddingEncoder | None = None
 
-        self._reference_window: list[tuple[int, int]] = []
-        self._current_window: list[tuple[int, int]] = []
-        self._total_items_in_current_window = 0
+        self._sample_left_until_detection = (
+            config.detection_interval_data_points
+        )  # allows to detect drift in a fixed interval
+        self._windows = _setup_detection_window_manager(config.windowing_strategy)
         self._triggered_once = False
 
         self.evidently_detector = EvidentlyDriftDetector(config.metrics)
         self.alibi_detector = AlibiDriftDetector(config.metrics)
 
+        # Every decision policy wraps one metric and is responsible for making decisions based on the metric's results
+        # and the metric's range of distance values
+        self.decision_policies = _setup_decision_engines(config)
+
+        # list of reference windows for each warmup interval
+        self.warmup_intervals: list[list[tuple[int, int]]] = []
```

Comment:
I guess it's still unclear to me here why we can have multiple warmup intervals

Reply:
Assume we have 5 warmup intervals. Then we first collect the first 5 references windows and do the triggering decision according to a simple time/amount based warmup policy. After finishing the warmup we have collected a couple of data intervals and run the drift detection with the most recent current window and all of the collected references window. Each of the reference windows generates a distance value together with the most recent current window. With this series of distances we can calibrate the dynamic threshold decision policy. Does that make sense?

Reply:
Okay, I think my first level of confusion stemmed from the fact that I believed this means we somehow have more than 1 warmup phase. But I understand now this is about waiting for `n` windows until warmup ends.

What I am now confused about is why we need to cache the reference windows here. Couldn't we, on the fly during warmup, do the drift detection (without issuing a trigger), and remember the distance values? In my understanding, if we have a warmup duration of 5, we'd need to cache 5 distance metrics. I don't understand yet why we need to cache the windows instead. You say that you run the detection with the current window for all reference windows, but why? We want to calibrate the dynamic threshold. For this, shouldn't we use the current data as it will be later on? (Sorry, hard to articulate what I mean - basically during the actual runtime after warmup, we use current/reference windows, but they are not 5 windows apart. This might influence the magnitude of the distance values, since during warmup if we compare a very old reference window to a new current window that might lead to a larger distance value.)

Reply:
Yes, this influences the magnitude, and that's exactly why we do it. After the warmup phase, we are on a "diagonal" element of the distance matrix. However, if we also use the corresponding reference/current window pairs from every interval in the warmup phase, we would see only close-to-zero if not only `0` distances, making the calibration useless. In my offline exploration, those policies then only took the diagonal walk. I treat this cache of windows more as a implementational detail that leads to a successful calibration.

Reply:
Okay, I see. I guess you're going to write about that in your thesis, if you already have a text statement prepared about this, would be good to explain this in a comment or the markdown documentation i suppose, because that was not obvious to me